aesd-char-driver/main.c

```diff
@@ -0,0 +1,287 @@
+{
+    struct aesd_dev *dev; /* Pointer to the aesd_dev struct*/
+    
+    PDEBUG("open");
+    
+	dev = container_of(inode->i_cdev, struct aesd_dev, cdev);   //The address of the container structure where cdev is located is returned and stored in dev
+	filp->private_data = dev; //THe open syscall sets filp to NULL before calling the open method to the driver. By assigning addr, other modules can use it. 
+    /**
+     * TODO: handle open
+     */
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    return 0;
+}
+
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    size_t offset_pos;
+    int kernel_buffer_count = 0;
+    struct aesd_dev *dev = filp->private_data;
+    struct aesd_buffer_entry *kernel_buff;
+
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+
+    mutex_lock(&aesd_device.lock);  //Kernel lock primitive
+
+    kernel_buff = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->aesd_circular_buffer, *f_pos, &offset_pos);
+
+    if (kernel_buff == NULL) //if address was not returned. 
+    {      
+        *f_pos = 0; 
+        goto clean; //goto is not frowned upon in kernel programming, use it to jump to cleanup steps. 
+    }
+    
+    if ((kernel_buff->size - offset_pos) < count) 
+    {
+        *f_pos += (kernel_buff->size - offset_pos);    //Based on the returned offset and size, update f_pos. 
+        kernel_buffer_count = kernel_buff->size - offset_pos;
+    } 
+    
+    else 
+    {
+        *f_pos += count;
+        kernel_buffer_count = count;
+    }
+
+    if (copy_to_user(buf, kernel_buff->buffptr+offset_pos, kernel_buffer_count)) //copying to buf, which is a userspace buffer.
+    {      
+		retval = -EFAULT;
+		goto clean;
+	}
+
+    retval = kernel_buffer_count;     
+    /**
+     * TODO: handle read
+     */
+    clean: mutex_unlock(&aesd_device.lock);
+
+    PDEBUG("Return Value %ld", retval);
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0; // Return Value
+    
+    char *temp_buffer; 
+    int i; // Loop iterator
+    int string_end_flag = 0; 
+    int temp_iterator = 0; 
+    int temp_counter = 0;
+    struct aesd_buffer_entry aesd_buffer_write_entry; 
+    struct aesd_dev *dev;
+
+    char *ret_ptr;
+
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+
+    dev = filp->private_data;
+
+    mutex_lock(&aesd_device.lock);  //kmalloc may sleep, use mutex to lock. 
+
+   
+    temp_buffer = (char *)kmalloc(count, GFP_KERNEL);   
+
+    if (temp_buffer == NULL) 
+    {
+        retval = -ENOMEM;
+        goto exit_clean;
+    }
+        
+    // Copying into the kernel buffer from buf
+    if (copy_from_user(temp_buffer, buf, count)) {
+        retval = -EFAULT;
+		goto exit_clean;
```

Comment:
@suja8742 what happens to `temp_buffer` in this failure path?

Reply:
Forgot to free, handled it in current implementation.